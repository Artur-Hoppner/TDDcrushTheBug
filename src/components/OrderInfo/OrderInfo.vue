<template>
  <section class="orderInfo">
    <h1 class="orderNumber">
      Thank you for your order! Your order number is:
      <span>{{ getOrderInfo.orderNumber }}</span>
    </h1>
    <br />
    <h3 class="orderList" :key="item.id" v-for="item in getOrderItems">
      Product: <span>{{ item.title }}</span>
      <br />
      <img
        class="product-pictures"
        :src="`${publicPath}assets/${item.tag}.png`"
        :alt="item.title"
      />
      <br />
      Quantity: <span>{{ item.quantity }}</span>
      <br />
      Cost: <span>{{ item.price }}:-</span>
    </h3>
    <br />
    <h1 class="orderCost">
      Total price: <span>{{ getOrderInfo.orderCost }}</span
      >:-
    </h1>
    <h2 class="deliveryETA">
      Expected delivery in: <span>{{ getOrderInfo.deliveryETA }}</span> days
    </h2>
    <img
      id="exterminator-thumbsup"
      src="@/assets/backgroundImages/exterminator-3.png"
      alt="exterminator thumbs up"
    />
  </section>
</template>
<script>
import { mapGetters } from 'vuex';

export default {
  name: 'OrderInfo',
  data() {
    return {
      publicPath: process.env.BASE_URL
    };
  },
  computed: {
    ...mapGetters(['getOrderInfo', 'getOrderItems'])
  }
};
</script>

<style lang="scss" scoped>
@import url('https://fonts.googleapis.com/css2?family=Montserrat:ital@1&family=Raleway&display=swap%27');

* {
  // box-sizing: border-box;
}

.orderInfo {
  font-family: 'Raleway', sans-serif;
  padding-top: 2rem;
}

span {
  font-family: 'Montserrat', sans-serif;
  font: bolder;
}

.product-pictures {
  width: 6vw;
  height: 6vw;
}

.deliveryETA {
  margin-bottom: -15%;
}

#exterminator-thumbsup {
  width: 30vw;
  height: 30vw;
  margin-left: 0;
  margin-top: 0;
  margin-bottom: -0.25rem;
  margin-right: -60%;

}
//font-family: 'Montserrat', sans-serif;
//font-family: 'Raleway', sans-serif;
</style>
