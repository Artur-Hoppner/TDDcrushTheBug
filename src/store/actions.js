const actions = {
  addThisToCart(context, product) {
    context.commit('addToCart', product);
  },
  changeThisProductButtonToggle(context, boolean) {
    context.commit('changeProductButtonToggle', boolean);
  },
  removeThisFromCart(context, id) {
    context.commit('removeFromCart', id);
  },
  increaseThisQuantity(context, product) {
    context.commit('increaseQuantity', product);
  },
  decreaseThisQuantity(context, product) {
    context.commit('decreaseQuantity', product);
  },
  getByThisKeyword(context, keyWord) {
    context.commit('getByKeyword', keyWord);
  },
  displayThisEntireArray(context) {
    context.commit('displayEntireArray');
  }
};

export default actions;
