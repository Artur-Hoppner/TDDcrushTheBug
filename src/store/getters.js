//***********************/
//*** GLOBAL GETTERS ***/
//*********************/

const getters = {
  getProductButtonToggle: state => {
    return state.productButtonToggle;
  },
  getCartProducts: state => {
    return state.cart;
  },
  filteredByKeyword: state => {
    return state.filteredProducts;
  },
  getOrderInfo: state => {
    return state.orderInfo;
  },
  getOrderItems: state => {
    return state.orderInfo.orderList;
  },
  cartIconItems: state => {
    let total = 0;
    state.cart.forEach(cart => {
      total += cart.quantity;
    });
    return total;
  },
  showHamburger: state => {
    return state.isBurgerActive;
  }
};

export default getters;
