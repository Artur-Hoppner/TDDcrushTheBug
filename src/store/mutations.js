//*************************/
//*** GLOBAL MUTATIONS ***/
//***********************/

const mutations = {
  //*************************************/
  //*** CART ARRAY RELATED MUTATIONS ***/
  //***********************************/
  //Add product to cart, takes product object as argument
  addToCart(state, product) {
    //Check if cart already has existing item so no doublettes will be created, only update quantity & price
    let checkExisting = state.cart.filter(item => item.id == product.id);
    if (checkExisting.length >= 1) {
      let index = state.cart.findIndex(item => item.id === product.id);
      state.cart[index].quantity++;
      state.cart[index].price += product.price;
    } else {
      state.cart.push({
        id: product.id,
        tag: product.tag,
        title: product.title,
        desc: product.desc,
        price: product.price,
        quantity: 1
      });
    }
  },
  //Remove Product from cart using index
  removeFromCart(state, id) {
    let findIndex = state.cart.find(item => item.id == id);
    let index = state.cart.indexOf(findIndex);
    state.cart.splice(index, 1);
  },
  //Increase Product Quantity in cart
  increaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Increase Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity++;
    state.cart[index].price += adjustPrice.price;
  },
  //Decrease Product Quantity in cart
  decreaseQuantity(state, product) {
    //Check for default values of object in original allProducts array
    //Decrease Product quantity and price
    let adjustPrice = state.allProducts.find(item => item.id == product.id);
    let index = state.cart.findIndex(item => item.id === product.id);
    state.cart[index].quantity--;
    state.cart[index].price -= adjustPrice.price;
    if (state.cart[index].quantity == 0) {
      state.cart.splice(index, 1);
    }
  },
  //Create order object based on state.cart
  checkoutCreateOrder(state) {
    if (state.cart != '') {
      //Randomize ordernumber and calculate total cost
      let newOrderNumber = Math.floor(Math.random() * 100000 + 1);
      let totalCost = 0;
      for (let i = 0; i < state.cart.length; i++) {
        totalCost += state.cart[i].price;
      }
      //Create Order object
      state.orderInfo = {
        orderNumber: newOrderNumber,
        orderCost: totalCost,
        deliveryETA: '3',
        orderList: state.cart
      };
    } else {
      return false;
    }
  },
  //*************************************************/
  //*** FILTEREDPRODUCTS ARRAY RELATED MUTATIONS ***/
  //***********************************************/
  //Take input searchword as argument, then use it to filter through allproducts array and match all products including the searchword
  getByKeyword(state, keyWord) {
    const search = keyWord.toLowerCase();
    if (!search.length) {
      state.filteredProducts = state.allProducts;
    } else {
      let foundProducts = [];
      foundProducts = state.allProducts.filter(item => item.tag.match(search));
      state.filteredProducts = foundProducts;
    }
  },
  //Reset the displayArray so it shows all products
  displayEntireArray(state) {
    state.filteredProducts = state.allProducts;
  },
  //*********************************/
  //*** TOGGLE RELATED MUTATIONS ***/
  //*******************************/
  //Change the toggle which decides if ProductItem shows AddToCart button or Increment, Decrement & RemoveFromCart depending on if we are viewing cart or shopping page
  changeProductButtonToggle(state, boolean) {
    state.productButtonToggle = boolean;
  },
  changeCart(state) {
    state.showCart = !state.showCart;
  },
  toggle(state) {
    state.isBurgerActive = !state.isBurgerActive
  }
};

export default mutations;
