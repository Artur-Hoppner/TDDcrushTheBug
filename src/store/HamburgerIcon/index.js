import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,

  state: {
    isBurgerActive: false
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
