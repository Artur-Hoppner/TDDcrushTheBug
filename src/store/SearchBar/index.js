import mutations from './mutations.js';
import getters from './getters.js';
import actions from './actions.js';

const state = {
  namespaced: true,
  //TODO: fix placeholder array with actual products!
  state: {
    allProducts: [
      'cat',
      'dog',
      'parrot',
      'goldfish',
      'horse',
      'elephant',
      'ant',
      'snake',
      'crow',
      'cow',
      'eagle',
      'rhinoceros',
      'chimpanzee'
    ],
    filteredProducts: []
  },
  mutations: mutations,
  getters: getters,
  actions: actions
};

export default state;
