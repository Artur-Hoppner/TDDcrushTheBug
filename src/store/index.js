import Vue from 'vue';
import Vuex from 'vuex';
import searchbar from './SearchBar/index.js';
import actions from './actions.js';
import mutations from './mutations.js';
import getters from './SearchBar/getters.js';
import hamburgerIcon from './HamburgerIcon/index.js'

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
    allProducts: [],
    cart: [],
    searchedProducts: []
  },
  actions: actions,
  mutations: mutations,
  getters: getters,
  modules: {
    searchbar: searchbar,
    hamburgerIcon: hamburgerIcon
  }
});
