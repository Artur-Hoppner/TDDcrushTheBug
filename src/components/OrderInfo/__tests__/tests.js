import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import store from '../../../store/index.js';
import OrderInfo from '@/components/OrderInfo/OrderInfo.vue';

//create a local instance of our vue
const localVue = createLocalVue();
//make our instance use vuex
localVue.use(Vuex);

describe('User is sent to view Order after checkout, expecting orderinfo to be displayed', () => {
  test('Test so that everything renders correctly', async () => {
    //Mount component with store & localvue, take snapshot and check so it renders correctly
    const wrapper = shallowMount(OrderInfo, { store, localVue });
    expect(wrapper.element).toMatchSnapshot();
  });

  test('Test that rendered output is representative of what is actually in the order Object', async () => {
    //Create an order
    const order = {
      orderNumber: 843888,
      orderCost: 218,
      deliveryETA: '3 working days',
      orderList: [
        {
          id: '10',
          tag: 'zapper',
          title: 'Regular zapper',
          desc: 'It tickles a little bit',
          price: 98
        },
        {
          id: '09',
          tag: 'mosquitozapper',
          title: 'Mosquito Zapper',
          desc: 'Designed especially for mosquitos, whatever that means',
          price: 120
        }
      ]
    };
    //Mock getters to return order that is created above
    const getters = {
      getOrderInfo() {
        return order;
      },
      getOrderItems() {
        return order.orderList;
      }
    };

    const store = new Vuex.Store({ getters });
    const wrapper = shallowMount(OrderInfo, { store, localVue });

    //Find all output elements
    const orderNumber = wrapper.find('.orderNumber');
    const orderCost = wrapper.find('.orderCost');
    const deliveryETA = wrapper.find('.deliveryETA');
    const orderList = wrapper.findAll('.orderList');

    //Test that output is representative of what order contains
    expect(orderNumber.text()).toContain(order.orderNumber);
    expect(orderCost.text()).toContain(order.orderCost);
    expect(deliveryETA.text()).toContain(order.deliveryETA);
    expect(orderList.length).toBeGreaterThan(0);
  });
});
