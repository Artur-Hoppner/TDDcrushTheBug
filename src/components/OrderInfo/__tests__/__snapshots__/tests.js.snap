// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User is sent to view Order after checkout, expecting orderinfo to be displayed Test so that everything renders correctly 1`] = `
<section
  class="orderInfo"
>
  <h1
    class="orderNumber"
  >
    
    Thank you for your order! Your order number is:
    
    <span>
      
    </span>
  </h1>
   
  <br />
    
  <br />
   
  <h1
    class="orderCost"
  >
    
    Total price: 
    <span>
      
    </span>
    :-
  
  </h1>
   
  <h2
    class="deliveryETA"
  >
    
    Expected delivery in: 
    <span>
      
    </span>
     days
  
  </h2>
   
  <img
    alt="exterminator thumbs up"
    id="exterminator-thumbsup"
    src="@/assets/backgroundImages/exterminator-3.png"
  />
</section>
`;
